SwiftCompilerSources/Sources/Optimizer/Utilities/OwnershipLiveness.swift

```diff
@@ -0,0 +1,977 @@
+///
+/// `definingValue`'s lifetime must already complete on all paths
+/// (a.k.a linear). Only lifetime-ending operations generate liveness.
+///
+/// `definingValue` dominates the range. Forwarding and phi uses do
+/// not extend the lifetime.
+///
+/// This is the simplest OSSA liveness analysis. It assumes that OSSA
+/// lifetime completion has already run on `definingValue`, and it
+/// cannot fix OSSA lifetimes after a transformation.
+func computeLinearLiveness(for definingValue: Value, _ context: Context)
+  -> InstructionRange {
+
+  assert(definingValue.ownership == .owned
+    || BeginBorrowValue(definingValue) != nil,
+    "value must define an OSSA lifetime")
+
+  // InstructionRange cannot directly represent the beginning of the block
+  // so we fake it with getRepresentativeInstruction().
+  var range = InstructionRange(for: definingValue, context)
+
+  // Compute liveness.
+  definingValue.uses.endingLifetime.forEach {
+    range.insert($0.instruction)
+  }
+  return range
+}
+
+/// Indicate whether OwnershipUseVisitor is visiting a use of the
+/// outer OSSA lifetime or within an inner borrow scope (reborrow).
+enum IsInnerLifetime {
+case outerLifetime
+case innerLifetime
+}
+
+typealias InnerScopeHandler = (Value) -> WalkResult
+
+/// Compute liveness and return a range, which the caller must deinitialize.
+///
+/// An OSSA lifetime begins with a single "defining" value, which
+/// must be owned, or must begin a borrow scope.
+///
+/// - Liveness does not extend beyond lifetime-ending operations
+/// (a.k.a. affine lifetimes).
+///
+/// - The definition dominates all use points.
+///
+/// - Does not assume the current lifetime is complete, but does
+/// assume any inner scopes are complete. Use `innerScopeHandler` to
+/// complete them or bail-out.
+func computeInteriorLiveness(for definingValue: Value,
+  _ context: FunctionPassContext,
+  innerScopeHandler: InnerScopeHandler? = nil) -> InstructionRange {
+
+  assert(definingValue.ownership == .owned
+    || BeginBorrowValue(definingValue) != nil,
+    "value must define an OSSA lifetime")
+
+  var range = InstructionRange(for: definingValue, context)
+
+  var visitor = InteriorUseVisitor(definingValue: definingValue, context) {
+    range.insert($0.instruction)
+    return .continueWalk
+  }
+  defer { visitor.deinitialize() }
+  visitor.innerScopeHandler = innerScopeHandler
+  let success = visitor.visitUses()
+  switch visitor.pointerStatus {
+  case .nonEscaping:
+    break
+  case let .escaping(operand):
+    fatalError("""
+                 check findPointerEscape() before computing interior liveness.
+                 Pointer escape: \(operand.instruction)
+                 """)
+  case let .unknown(operand):
+    fatalError("Unrecognized SIL address user \(operand.instruction)")
+  }
+  assert(success == .continueWalk, "our visitor never fails")
+  assert(visitor.unenclosedPhis.isEmpty, "missing adjacent phis")
+  return range
+}
+
```

Comment:
Forwarding is a simple, well-defined concept that is independent from any other OSSA concepts. The analysis of borrow scopes relies on forwarding. But it's often sufficient to just check ForwardingInstruction, which is much simpler than conforming to a walker utility. Sometimes LifetimeDependence also needs to know about forwarding, but it is also a separate concept.

This is an example of why composition is better than inheritance. Each utility does exactly one thing, which is very simple and well-defined but can be used in many different contexts.

The biggest problem with our previous utilities is that the don't cleanly model well-defined concepts. We merge utilities to reduce lines of code and in the process generate overwhelming complexity. A single utility that means different things in different contexts is the worst kind of complexity. It's like using a type for something that is not that type just because it shares a representation.